pkg/pillar/cmd/zedrouter/appcontainer.go

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2020 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// Process input in the form of a collection of AppNetworkConfig structs
+// from zedmanager and zedagent. Publish the status as AppNetworkStatus.
+// Produce the updated configlets (for radvd, dnsmasq, ip*tables, lisp.config,
+// ipset, ip link/addr/route configuration) based on that and apply those
+// configlets.
+
+package zedrouter
+
+import (
+	"strings"
+	"time"
+
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	log "github.com/sirupsen/logrus"
+)
+
+// check if we need to launch the goroutine to collect App container stats
+func appCheckStatsCollect(ctx *zedrouterContext, config types.AppNetworkConfig,
+	status *types.AppNetworkStatus) {
+
+	oldIPAddr := status.GetStatsIPAddr
+	status.GetStatsIPAddr = config.GetStatsIPAddr
+	if strings.Compare(config.GetStatsIPAddr, oldIPAddr) != 0 {
+		log.Infof("appCheckStatsCollect: config ip %s, status ip %s", config.GetStatsIPAddr, oldIPAddr)
+		if oldIPAddr == "" && config.GetStatsIPAddr != "" {
+			ctx.appStatsMutex.Lock()
+			publishAppNetworkStatus(ctx, status)
+			if !ctx.appCollectStatsRunning {
+				ctx.appStatsMutex.Unlock()
+				go appStatsCollect(ctx)
```

Comment:
This can easily start multiple instances of the goroutine. I think you need to set appCollectStatsRunning under the mutex to avoid that.

Also, might want to encapsulate this under a ensureStatsCollectRunning(ctx) to hide the logic.